elementary/monitor/alerts/group_of_alerts.py

```diff
@@ -0,0 +1,337 @@
+from dataclasses import dataclass
+from datetime import datetime
+from enum import Enum
+from typing import Dict, List
+
+from elementary.monitor.alerts.alert import Alert, SlackAlertMessageBuilder
+from elementary.monitor.alerts.model import ModelAlert
+from elementary.monitor.fetchers.alerts.normalized_alert import CHANNEL_KEY
+from elementary.utils.json_utils import try_load_json
+from elementary.utils.models import alert_to_concise_name, get_shortened_model_name
+
+DATETIME_FORMAT = "%Y-%m-%d %H:%M:%S"
+
+
+class GroupingType(Enum):
+    BY_ALERT = "alert"
+    BY_TABLE = "table"
+    ALL = "all"
+
+
+@dataclass(
+    frozen=True, eq=True
+)  # frozen+eq defined so we can use it as a dict key. Also, it's all Strings
+class NotificationComponent:
+    name_in_summary: str
+    empty_section_content: str
+
+
+@dataclass(frozen=True, eq=True)
+class AlertGroupComponent(NotificationComponent):
+    emoji_in_summary: str
+    name_in_full: str
+    emoji_in_full: str
+
+
+ErrorComponent = AlertGroupComponent(
+    name_in_summary="Errors",
+    emoji_in_summary="exclamation",
+    name_in_full="Error",
+    emoji_in_full="exclamation",
+    empty_section_content="No Errors",
+)
+
+WarningComponent = AlertGroupComponent(
+    name_in_summary="Warning",
+    emoji_in_summary="warning",
+    name_in_full="Warning",
+    emoji_in_full="warning",
+    empty_section_content="No Warnings",
+)
+
+FailureComponent = AlertGroupComponent(
+    name_in_summary="Failed",
+    emoji_in_summary="small_red_triangle",
+    name_in_full="Failed tests",
+    emoji_in_full="X",
+    empty_section_content="No Failures",
+)
+
+
+TagsComponent = NotificationComponent(
+    name_in_summary="Tags", empty_section_content="No Tags"
+)
+OwnersComponent = NotificationComponent(
+    name_in_summary="Owners", empty_section_content="No Owners"
+)
+SubsComponent = NotificationComponent(
+    name_in_summary="Subscribers", empty_section_content="No Subscribers"
+)
+
+
+class GroupOfAlerts(SlackAlertMessageBuilder):
```

Comment:
I think that this class shouldn't inheritance from SlackAlertMessageBuilder.
The fact that it has to_slack methods means that this is a class that uses a SlackMessageBuilder but not an instance of SlackMEssageBuilder.

I think that having this class which processes the data and using a SlackGroupAlertMessageBuilder (in similar to how Alert class works) will make sense:
This class will format the data and gather the tags, owners, subscribers, etc that the SlackMessageBuilder needed, and use it to generate the Slack message.
That way you can relay on a single format of data without duplicates (for tags, owners, subscribers for example) at the SlackMessageBuilder and only focus on the generation of the message itself.
WDYT?

Reply:
accepted